### PR TITLE
System instant

### PR DESCRIPTION
**Pull request by folkertdev**
use the system instant instead of a wrapped ntp timestamp

### CODE REVIEW

File: ntp-daemon/src/peer.rs [hunk 1]
```diff
@@ -59,13 +65,16 @@ pub async fn start_peer<A: ToSocketAddrs, C: 'static + NtpClock + Send>(
                         .as_mut()
                         .reset(Instant::now() + poll_interval.as_system_duration());
 
-                    // TODO: Figure out proper error behaviour here
-                    let ntp_instant = NtpInstant::from_ntp_timestamp(clock.now().unwrap());
+                    let ntp_instant = NtpInstant::now();
                     let packet = peer.generate_poll_message(ntp_instant);
+
+                    // TODO error handling here; what if we cannot determine the origin timestamp?
+                    last_send_timestamp = clock.now().ok();
```
Comment (davidv1992) [inline ntp-daemon/src/peer.rs:72]: If we can't properly stamp a send timestamp, that pretty much inhibits any synchronization, so in my view crashing out here with an expect is acceptable. The only other alternative I see is to log the error and not send the packet, but given the type of causes for the clock failing I don't really see this doing much in terms of resilience (it is pretty much guaranteed to always fail if it fails once)

File: ntp-proto/src/filter.rs [hunk 2]
```diff
@@ -118,7 +126,8 @@ impl LastMeasurements {
         system_precision: NtpDuration,
         frequency_tolerance: FrequencyTolerance,
     ) -> Option<(PeerStatistics, NtpInstant)> {
-        let dispersion_correction = (new_tuple.time - peer_time) * frequency_tolerance;
+        let dispersion_correction =
+            NtpInstant::abs_diff(new_tuple.time, peer_time) * frequency_tolerance;
```
Comment (davidv1992) [inline ntp-proto/src/filter.rs:130]: This is a mild change of behaviour compared to the old situation, although new_tuple.time should never be before peer_time. However, perhaps we should assert this? (at least in debug mode)

File: ntp-proto/src/time_types.rs [hunk 3]
```diff
@@ -1,47 +1,50 @@
...
+    pub fn now() -> Self {
         Self {
-            timestamp: NtpTimestamp::from_bits(bits),
+            instant: Instant::now(),
         }
     }
 
-    pub(crate) const fn to_bits(self) -> [u8; 8] {
-        self.timestamp.to_bits()
+    // used to populate the T1 and T3 fields to ensure the server is doing something sensible
```
Comment (davidv1992) [inline ntp-proto/src/time_types.rs:19]: I think this is (or at least should) only ever be used for the T1 timestamp. T3 isn't even ours to control as that is the timestamp the server indicated it sent its packet at.

File: ntp-proto/src/time_types.rs [hunk 4]
```diff
@@ -1,47 +1,50 @@
...
+    type Output = NtpInstant;
 
-    fn sub(self, rhs: Self) -> Self::Output {
-        self.timestamp - rhs.timestamp
+    fn add(mut self, rhs: Duration) -> Self::Output {
+        self.instant += rhs;
+
+        self
     }
 }
```
Comment (davidv1992) [inline ntp-proto/src/time_types.rs:49]: I think we only ever need this in testing code, so we should probably mark it with the appropriate cfg marker.

**Review by davidv1992 — COMMENTED** [review]
Very nice PR, left a few minor comments.

File: ntp-proto/src/time_types.rs (hunk 3 above)
Comment (folkertdev) [reply ntp-proto/src/time_types.rs:19]: exactly, so we populate it, and then we check that the server has overwritten what we wrote

Comment (davidv1992) [reply ntp-proto/src/time_types.rs:19]: Sorry, but I think I am missing the point here. The server doesn't write our memory right, so I am not getting what you are implying here

**Pull request merged**